accepted/cranelift-isel-isle-peepmatic.md

```diff
@@ -0,0 +1,926 @@
+# Summary
+
+This RFC proposes a new instruction selection DSL for Cranelift, based
+on initial discussions begun with a pre-RFC (bytecodealliance/rfcs#13).
+
+This RFC proposes the name "ISLE" (Instruction Selection/Lowering
+Expressions), pronounced like the word for a small island (silent
+'s'), which is conveniently also an anagram of "isel".
+
+This DSL combines aspects of an earlier prototype design that the
+author has developed with ideas from term-rewriting systems, and in
+particular plans to implement on top of Peepmatic's automata-based
+compilation strategy with a new Rust-generating backend.
+
+The basic principles, in terms introduced/framed by the pre-RFC linked
+above, are:
+
+- Expression/value-based design with term-rewriting-like semantics.
+- Interaction with existing system, including rich query APIs, via
+  "virtual nodes" defined by external extractor functions.
+- Typed terms.
+- Rule application order guided by explicit priorities, but
+  semantically undefined (hence flexible).
+- Single-pass lowering semantics via unrolling.
+
+We will dive into each of these principles in more detail below.
+
+# Intro: Term-Rewriting System
+
+The basic concept behind the sort of system that this DSL embodies is
+that of *term-rewriting*. In brief, this means that the input and
+output of the system are trees (or more general graphs) of *terms*
+(symbolic nodes with arguments), and the system transforms input to
+output by applying a set of *rewrite rules*. Each rule matches on some
+*pattern* -- for example, a term of a particular type with an input
+that matches some sub-pattern -- and replaces it with a new tree of
+terms.
+
+This is a natural abstraction for transforming code. If we define a
+term view of the IR on the input side and machine instructions on the
+output side, such that a term represents an operator that consumes
+values and produces a value, then the rewrite rules just define
+*equivalences*. These rules are easier to validate than other sorts of
+compiler-backend implementations because "all" one has to do is to
+show that the two sides of the equivalence are equal, separately for
+each rule (i.e., the validation is modular).
+
+An example of a rewrite rule might be:
+
+```plain
+
+    (rule (Iadd a b)
+          (X86Add a b))
```

Comment:
A suggestion for an alternative syntax to consider:

```
(Iadd a b) => (X86Add a b);
```

Reply:
I'm generally hoping to stay true to S-expression form here (so the toplevel forms/clauses are `(KEYWORD ...)`) but `=>` as the keyword (`(=> (A x) (B x))`), which is what Peepmatic does, could work. Others please let us know your favorite bikeshed color below!

Reply:
One more suggestion: Please enforce a specific formatting of the S-expressions on CI in rustfmt style. That is a 1-to-1 map from parsed AST to the resulting string. Not like eg clang-format which tries to preserve newlines and things like that and only normalizes things like spaces.